File: financial_ai_sdk.ts
/**
 * FinancialAI SDK - Cliente TypeScript/JavaScript
 * 
 * Uso:
 *   const ai = new FinancialAI({ apiUrl: 'http://localhost:5000', token: 'YOUR_TOKEN' });
 *   const result = await ai.fullAnalysis();
 */

export interface AnalysisOptions {
  includeCharts?: boolean;
  cacheTTL?: number;
  timeout?: number;
}

export interface PredictionData {
  categoria?: Record<string, number>;
  mensual?: Record<string, number>;
  anomalias?: {
    detectadas: number;
    porcentaje: number;
  };
  modelos?: Record<string, number>;
  [key: string]: any;
}

export interface StatisticsData {
  correlaciones?: Record<string, number>;
  tendencias?: {
    promedio_diario: number;
    máximo: number;
    mínimo: number;
  };
  clustering?: {
    grupos: number;
    varianza: number;
  };
  [key: string]: any;
}

export interface SavingsData {
  goals?: Array<{
    nombre: string;
    meta: number;
    prioridad: string;
  }>;
  tips?: string[];
  budget_alerts?: {
    limite_recomendado: number;
    limite_maximo: number;
  };
  health_score?: number;
  [key: string]: any;
}

export interface ChartsData {
  heatmap?: Record<string, any>;
  sankey?: Record<string, any>;
  dashboard?: {
    total_gastos: number;
    gastos_promedio: number;
    categorias: number;
  };
  [key: string]: any;
}

export interface AnalysisResult {
  success: boolean;
  data: {
    usuario_id: string;
    prediccion?: PredictionData;
    estadisticas?: StatisticsData;
    ahorro?: SavingsData;
    graficos?: ChartsData;
  };
  meta: {
    processing_time_ms: number;
    timestamp: string;
    cache_hit: boolean;
    nivel: 'quick' | 'full';
    modelos_utilizados?: string[];
  };
  errors?: string[];
}

export interface JobResult {
  success: boolean;
  data: {
    job_id: string;
    usuario_id: string;
    status: 'queued' | 'processing' | 'completed' | 'error';
    resultado?: any;
    error?: string;
    timestamp?: string;
  };
}

export class FinancialAIError extends Error {
  constructor(message: string, public statusCode?: number) {
    super(message);
    this.name = 'FinancialAIError';
  }
}

export class FinancialAI {
  private apiUrl: string;
  private token: string;
  private timeout: number;
  private cache: Map<string, { data: any; timestamp: number }>;
  private cacheTTL: number;

  constructor(options: {
    apiUrl: string;
    token: string;
    timeout?: number;
    cacheTTL?: number;
  }) {
    this.apiUrl = options.apiUrl.replace(/\/$/, '');
    this.token = options.token;
    this.timeout = options.timeout || 30000;
    this.cacheTTL = options.cacheTTL || 300000; // 5 minutos por defecto
    this.cache = new Map();
  }

  /**
   * Realizar request a la API
   */
  private async request<T>(
    method: string,
    endpoint: string,
    body?: any
  ): Promise<T> {
    const url = `${this.apiUrl}${endpoint}`;
    const cacheKey = `${method}:${endpoint}`;

    // Verificar caché
    if (method === 'GET') {
      const cached = this.cache.get(cacheKey);
      if (cached && Date.now() - cached.timestamp < this.cacheTTL) {
        return cached.data;
      }
    }

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(url, {
        method,
        headers: {
          Authorization: `Bearer ${this.token}`,
          'Content-Type': 'application/json',
        },
        body: body ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });

      if (!response.ok) {
        throw new FinancialAIError(
          `API Error: ${response.statusText}`,
          response.status
        );
      }

      const data = await response.json();

      if (!data.success) {
        throw new FinancialAIError(data.error || 'Unknown error');
      }

      // Guardar en caché
      if (method === 'GET') {
        this.cache.set(cacheKey, { data, timestamp: Date.now() });
      }

      return data;
    } catch (error) {
      if (error instanceof FinancialAIError) {
        throw error;
      }
      if (error instanceof TypeError) {
        throw new FinancialAIError('Network error or API unreachable');
      }
      throw new FinancialAIError(
        error instanceof Error ? error.message : 'Unknown error'
      );
    } finally {
      clearTimeout(timeoutId);
    }
  }

  /**
   * Análisis rápido (< 2 segundos)
   */
  async quickAnalysis(): Promise<AnalysisResult> {
    return this.request<AnalysisResult>(
      'POST',
      '/api/v2/analysis/quick'
    );
  }

  /**
   * Análisis completo (2-5 segundos)
   */
  async fullAnalysis(): Promise<AnalysisResult> {
    return this.request<AnalysisResult>(
      'POST',
      '/api/v2/analysis/full'
    );
  }

  /**
   * Encolar análisis profundo (asíncrono)
   */
  async queueAnalysis(): Promise<JobResult> {
    return this.request<JobResult>(
      'POST',
      '/api/v2/analysis/queue'
    );
  }

  /**
   * Verificar estado de un job
   */
  async checkJob(jobId: string): Promise<JobResult> {
    return this.request<JobResult>(
      'GET',
      `/api/v2/analysis/job/${jobId}`
    );
  }

  /**
   * Esperar a que un job se complete
   */
  async waitForJob(
    jobId: string,
    options: { timeout?: number; pollInterval?: number } = {}
  ): Promise<any> {
    const timeout = options.timeout || 120000; // 2 minutos
    const pollInterval = options.pollInterval || 2000; // 2 segundos
    const startTime = Date.now();

    while (Date.now() - startTime < timeout) {
      const result = await this.checkJob(jobId);
      const status = result.data.status;

      if (status === 'completed') {
        return result.data.resultado;
      }

      if (status === 'error') {
        throw new FinancialAIError(
          result.data.error || 'Job failed'
        );
      }

      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }

    throw new FinancialAIError(`Timeout waiting for job ${jobId}`);
  }

  /**
   * Ejecutar análisis completo con espera
   */
  async fullAnalysisAsync(
    options: { timeout?: number; pollInterval?: number } = {}
  ): Promise<any> {
    const queueResult = await this.queueAnalysis();
    const jobId = queueResult.data.job_id;
    return this.waitForJob(jobId, options);
  }

  /**
   * Cancelar caché
   */
  clearCache(): void {
    this.cache.clear();
  }

  /**
   * Obtener estadísticas del caché
   */
  getCacheStats(): { size: number; keys: string[] } {
    return {
      size: this.cache.size,
      keys: Array.from(this.cache.keys()),
    };
  }
}

/**
 * Hook React para usar FinancialAI
 */
export function useFinancialAI(token: string, apiUrl: string = 'http://localhost:5000') {
  const [ai] = React.useState(() => new FinancialAI({ apiUrl, token }));

  const [state, setState] = React.useState<{
    data: AnalysisResult | null;
    loading: boolean;
    error: FinancialAIError | null;
  }>({
    data: null,
    loading: false,
    error: null,
  });

  const quickAnalysis = React.useCallback(async () => {
    setState({ data: null, loading: true, error: null });
    try {
      const result = await ai.quickAnalysis();
      setState({ data: result, loading: false, error: null });
      return result;
    } catch (error) {
      const err = error instanceof FinancialAIError ? error : new FinancialAIError(String(error));
      setState({ data: null, loading: false, error: err });
      throw err;
    }
  }, [ai]);

  const fullAnalysis = React.useCallback(async () => {
    setState({ data: null, loading: true, error: null });
    try {
      const result = await ai.fullAnalysis();
      setState({ data: result, loading: false, error: null });
      return result;
    } catch (error) {
      const err = error instanceof FinancialAIError ? error : new FinancialAIError(String(error));
      setState({ data: null, loading: false, error: err });
      throw err;
    }
  }, [ai]);

  return { ...state, quickAnalysis, fullAnalysis };
}

/**
 * Formatter para mostrar resultados
 */
export class AnalysisFormatter {
  static formatPredictionSummary(pred: PredictionData): string {
    const lines: string[] = [];
    lines.push('📊 PREDICCIONES');

    if (pred.categoria) {
      lines.push('\nGastos por categoría:');
      Object.entries(pred.categoria).forEach(([cat, amount]) => {
        lines.push(`  • ${cat}: $${(amount as number).toFixed(2)}`);
      });
    }

    if (pred.anomalias) {
      lines.push(`\nAnomalías: ${pred.anomalias.detectadas} (${pred.anomalias.porcentaje.toFixed(1)}%)`);
    }

    return lines.join('\n');
  }

  static formatSavingsSummary(savings: SavingsData): string {
    const lines: string[] = [];
    lines.push('💰 AHORRO');

    if (savings.health_score) {
      const score = savings.health_score;
      const status = score > 80 ? '⭐ Excelente' : score > 60 ? '🟡 Bueno' : '🔴 Necesita mejora';
      lines.push(`\nSalud financiera: ${score}% ${status}`);
    }

    if (savings.tips && savings.tips.length > 0) {
      lines.push('\nRecomendaciones:');
      savings.tips.forEach((tip) => {
        lines.push(`  • ${tip}`);
      });
    }

    return lines.join('\n');
  }

  static formatReport(result: AnalysisResult): string {
    const lines: string[] = [];

    if (result.data.prediccion) {
      lines.push(this.formatPredictionSummary(result.data.prediccion));
    }

    if (result.data.ahorro) {
      lines.push('\n' + this.formatSavingsSummary(result.data.ahorro));
    }

    lines.push(`\n⚡ Procesamiento: ${result.meta.processing_time_ms}ms`);

    return lines.join('\n');
  }
}

// ============================================================
// EJEMPLOS DE USO
// ============================================================

if (typeof window !== 'undefined') {
  // Ejemplo React
  const Example: React.FC = () => {
    const token = localStorage.getItem('token') || '';
    const { data, loading, error, fullAnalysis } = useFinancialAI(token);

    React.useEffect(() => {
      fullAnalysis();
    }, [fullAnalysis]);

    if (loading) return <div>Analizando...</div>;
    if (error) return <div>Error: {error.message}</div>;
    if (!data) return null;

    return (
      <div>
        <h1>Análisis Financiero</h1>
        <pre>{AnalysisFormatter.formatReport(data)}</pre>
        <p>Tiempo: {data.meta.processing_time_ms}ms</p>
      </div>
    );
  };

  // Ejemplo vanilla JavaScript
  async function exampleVanilla() {
    const ai = new FinancialAI({
      apiUrl: 'http://localhost:5000',
      token: 'your_token_here',
    });

    try {
      const result = await ai.fullAnalysis();
      console.log(AnalysisFormatter.formatReport(result));
    } catch (error) {
      console.error('Error:', error);
    }
  }
}

export default FinancialAI;

File: financial_ai_sdk.py
"""
SDK de FinancialAI - Cliente Python para API de Análisis Financiero

Uso simple:
    ai = FinancialAI(api_url="http://localhost:5000", token="YOUR_TOKEN")
    result = ai.full_analysis()
"""

import requests
import json
from typing import Dict, Optional, List
from datetime import datetime
import time


class FinancialAIError(Exception):
    """Excepción base del SDK"""
    pass


class FinancialAI:
    """Cliente Python para acceder a los análisis financieros de IA"""
    
    def __init__(self, api_url: str, token: str, timeout: int = 30):
        """
        Inicializar el cliente
        
        Args:
            api_url: URL base de la API (ej: http://localhost:5000)
            token: Token JWT para autenticación
            timeout: Timeout en segundos para las requests
        """
        self.api_url = api_url.rstrip('/')
        self.token = token
        self.timeout = timeout
        self.headers = {
            'Authorization': f'Bearer {token}',
            'Content-Type': 'application/json'
        }
    
    def _request(self, method: str, endpoint: str, **kwargs) -> Dict:
        """
        Realizar una request a la API
        
        Args:
            method: GET, POST, etc
            endpoint: Ruta del endpoint (ej: /api/v2/analysis/full)
            **kwargs: Argumentos adicionales para requests
        
        Returns:
            Dict con la respuesta JSON
        
        Raises:
            FinancialAIError si hay error
        """
        url = f"{self.api_url}{endpoint}"
        
        try:
            kwargs.setdefault('headers', self.headers)
            kwargs.setdefault('timeout', self.timeout)
            
            response = requests.request(method, url, **kwargs)
            response.raise_for_status()
            
            data = response.json()
            
            if not data.get('success', True):
                raise FinancialAIError(data.get('error', 'Error desconocido'))
            
            return data
        
        except requests.exceptions.ConnectionError:
            raise FinancialAIError(f"No se pudo conectar a {url}")
        except requests.exceptions.Timeout:
            raise FinancialAIError(f"Timeout en {url}")
        except requests.exceptions.HTTPError as e:
            raise FinancialAIError(f"Error HTTP {response.status_code}: {response.text}")
        except json.JSONDecodeError:
            raise FinancialAIError("Respuesta no válida del servidor")
    
    def quick_analysis(self) -> Dict:
        """
        Análisis rápido (< 2 segundos)
        
        Retorna:
            - prediccion: Predicciones rápidas
            - estadisticas: Estadísticas básicas
        
        Example:
            >>> ai = FinancialAI(api_url, token)
            >>> result = ai.quick_analysis()
            >>> print(result['data']['prediccion'])
        """
        return self._request('POST', '/api/v2/analysis/quick')
    
    def full_analysis(self) -> Dict:
        """
        Análisis completo (2-5 segundos)
        
        Retorna:
            - prediccion: Todas las predicciones
            - estadisticas: Análisis estadístico completo
            - ahorro: Recomendaciones de ahorro
            - graficos: Datos para gráficos
        
        Example:
            >>> result = ai.full_analysis()
            >>> prediction = result['data']['prediccion']
            >>> savings = result['data']['ahorro']
            >>> charts = result['data']['graficos']
        """
        return self._request('POST', '/api/v2/analysis/full')
    
    def queue_analysis(self, wait: bool = False, timeout: int = 60) -> Dict:
        """
        Encola un análisis profundo (asíncrono)
        
        Args:
            wait: Si es True, espera a que termine (con timeout)
            timeout: Tiempo máximo en segundos para esperar
        
        Returns:
            Si wait=False: {'job_id', 'status', 'estimated_time_seconds'}
            Si wait=True: Resultado completo del análisis
        
        Example:
            >>> # Asíncrono
            >>> result = ai.queue_analysis()
            >>> job_id = result['data']['job_id']
            >>> status = ai.check_job(job_id)
            
            >>> # Con espera
            >>> result = ai.queue_analysis(wait=True, timeout=120)
            >>> print(result['data']['resultado'])
        """
        response = self._request('POST', '/api/v2/analysis/queue')
        
        if not wait:
            return response
        
        job_id = response['data']['job_id']
        start_time = time.time()
        
        while time.time() - start_time < timeout:
            job_status = self.check_job(job_id)
            status = job_status['data']['status']
            
            if status == 'completed':
                return job_status
            elif status == 'error':
                raise FinancialAIError(job_status['data'].get('error', 'Error desconocido'))
            
            time.sleep(2)  # Esperar 2 segundos antes de reintentar
        
        raise FinancialAIError(f"Timeout esperando job {job_id}")
    
    def check_job(self, job_id: str) -> Dict:
        """
        Verificar el estado de un análisis en queue
        
        Args:
            job_id: ID del job retornado por queue_analysis()
        
        Returns:
            Estado del job y resultado si está completado
        
        Example:
            >>> job_id = "job_abc123"
            >>> status = ai.check_job(job_id)
            >>> print(status['data']['status'])  # 'queued', 'processing', 'completed'
        """
        return self._request('GET', f'/api/v2/analysis/job/{job_id}')
    
    def cancel_job(self, job_id: str) -> Dict:
        """
        Cancelar un análisis en queue
        
        Args:
            job_id: ID del job a cancelar
        """
        return self._request('DELETE', f'/api/v2/analysis/job/{job_id}')
    
    def predict_category(self) -> Dict:
        """Predicción por categoría (heredado, usa quick_analysis ahora)"""
        return self._request('GET', '/api/v2/predict-category')
    
    def predict_monthly(self) -> Dict:
        """Predicción mensual (heredado, usa quick_analysis ahora)"""
        return self._request('GET', '/api/v2/predict-monthly')
    
    def detect_anomalies(self) -> Dict:
        """Detección de anomalías (heredado, usa quick_analysis ahora)"""
        return self._request('GET', '/api/v2/detect-anomalies')
    
    def get_savings_tips(self) -> Dict:
        """Obtener tips de ahorro (heredado, usa full_analysis ahora)"""
        return self._request('GET', '/api/v2/savings/tips')
    
    def get_health_score(self) -> Dict:
        """Obtener puntuación financiera (heredado, usa full_analysis ahora)"""
        return self._request('GET', '/api/v2/savings/health-score')
    
    def get_charts_dashboard(self) -> Dict:
        """Obtener datos del dashboard (heredado, usa full_analysis ahora)"""
        return self._request('GET', '/api/v2/charts/dashboard')


class FinancialAIAnalyzer:
    """Analizador de resultados para FinancialAI"""
    
    def __init__(self, analysis_result: Dict):
        """
        Inicializar con resultado de análisis
        
        Args:
            analysis_result: Dict retornado por full_analysis()
        """
        self.data = analysis_result.get('data', {})
        self.meta = analysis_result.get('meta', {})
    
    def get_prediction_summary(self) -> str:
        """Resumen de predicciones en texto"""
        pred = self.data.get('prediccion', {})
        
        lines = ["📊 RESUMEN DE PREDICCIONES"]
        lines.append("-" * 40)
        
        if 'categoria' in pred:
            lines.append("Gastos por categoría (predicción):")
            for cat, amount in pred['categoria'].items():
                lines.append(f"  • {cat}: ${amount:.2f}")
        
        if 'anomalias' in pred:
            ano = pred['anomalias']
            lines.append(f"\nAnomalías detectadas: {ano.get('detectadas', 0)}")
            lines.append(f"Porcentaje: {ano.get('porcentaje', 0):.1f}%")
        
        return "\n".join(lines)
    
    def get_savings_summary(self) -> str:
        """Resumen de oportunidades de ahorro"""
        savings = self.data.get('ahorro', {})
        
        lines = ["💰 RESUMEN DE AHORRO"]
        lines.append("-" * 40)
        
        if 'health_score' in savings:
            score = savings['health_score']
            status = "⭐ Excelente" if score > 80 else "🟡 Bueno" if score > 60 else "🔴 Necesita mejora"
            lines.append(f"Salud financiera: {score}% {status}")
        
        if 'tips' in savings:
            lines.append("\nRecomendaciones:")
            for tip in savings['tips']:
                lines.append(f"  • {tip}")
        
        if 'goals' in savings:
            lines.append("\nMetas de ahorro:")
            for goal in savings['goals']:
                lines.append(f"  • {goal.get('nombre')}: ${goal.get('meta'):.2f}")
        
        return "\n".join(lines)
    
    def get_performance_stats(self) -> str:
        """Estadísticas de rendimiento de la consulta"""
        lines = ["⚡ RENDIMIENTO"]
        lines.append("-" * 40)
        lines.append(f"Tiempo de procesamiento: {self.meta.get('processing_time_ms', 0):.0f}ms")
        lines.append(f"Timestamp: {self.meta.get('timestamp', 'N/A')}")
        lines.append(f"Nivel de análisis: {self.meta.get('nivel', 'N/A')}")
        lines.append(f"Cache hit: {self.meta.get('cache_hit', False)}")
        
        return "\n".join(lines)
    
    def print_report(self):
        """Imprimir reporte completo"""
        print(self.get_prediction_summary())
        print("\n")
        print(self.get_savings_summary())
        print("\n")
        print(self.get_performance_stats())


# ============================================================
# EJEMPLOS DE USO
# ============================================================

if __name__ == "__main__":
    # Configuración
    API_URL = "http://localhost:5000"
    TOKEN = "your_jwt_token_here"
    
    # Crear cliente
    ai = FinancialAI(api_url=API_URL, token=TOKEN)
    
    print("=" * 60)
    print("🚀 FINANCIAL AI - SDK PYTHON")
    print("=" * 60)
    
    try:
        # Ejemplo 1: Análisis rápido
        print("\n1️⃣ ANÁLISIS RÁPIDO (< 2 segundos)")
        print("-" * 60)
        quick = ai.quick_analysis()
        analyzer = FinancialAIAnalyzer(quick)
        analyzer.print_report()
        
        # Ejemplo 2: Análisis completo
        print("\n2️⃣ ANÁLISIS COMPLETO (2-5 segundos)")
        print("-" * 60)
        full = ai.full_analysis()
        analyzer = FinancialAIAnalyzer(full)
        analyzer.print_report()
        
        # Ejemplo 3: Análisis asíncrono
        print("\n3️⃣ ANÁLISIS ASÍNCRONO (PROFUNDO)")
        print("-" * 60)
        
        # Iniciar análisis sin esperar
        result = ai.queue_analysis(wait=False)
        job_id = result['data']['job_id']
        print(f"Job encolado: {job_id}")
        print(f"Tiempo estimado: {result['data']['estimated_time_seconds']}s")
        
        # Verificar estado
        print("\nVerificando estado...")
        status = ai.check_job(job_id)
        print(f"Estado: {status['data']['status']}")
        
        # Esperar a que termine
        print("\nEsperando a que se complete...")
        completed = ai.queue_analysis(job_id=job_id, wait=True)
        print(f"✅ Análisis completado!")
        
    except FinancialAIError as e:
        print(f"❌ Error: {e}")
    
    print("\n" + "=" * 60)
    print("✅ EJEMPLOS COMPLETADOS")
    print("=" * 60)
